modules/module-mysql/src/replication/zongji/BinLogListener.ts

```diff
@@ -201,43 +279,192 @@ export class BinLogListener {
             evt.tableMap[evt.tableId]
           );
+          this.logger.info(
+            `Processed Update row event of ${evt.rows.length} rows for table: ${evt.tableMap[evt.tableId].tableName}`
+          );
           break;
         case zongji_utils.eventIsDeleteMutation(evt):
           await this.eventHandler.onDelete(evt.rows, evt.tableMap[evt.tableId]);
+          this.logger.info(
+            `Processed Delete row event of ${evt.rows.length} rows for table: ${evt.tableMap[evt.tableId].tableName}`
+          );
           break;
         case zongji_utils.eventIsXid(evt):
+          this.binLogPosition.offset = evt.nextPosition;
           const LSN = new common.ReplicatedGTID({
             raw_gtid: this.currentGTID!.raw,
-            position: {
-              filename: this.binLogPosition.filename,
-              offset: evt.nextPosition
-            }
+            position: this.binLogPosition
           }).comparable;
           await this.eventHandler.onCommit(LSN);
+          this.logger.info(`Processed Xid log event. Transaction LSN: ${LSN}.`);
           break;
       }
 
       this.queueMemoryUsage -= evt.size;
     };
   }
 
-  isQueueOverCapacity(): boolean {
-    return this.queueMemoryUsage >= this.queueMemoryLimit;
+  private async processQueryEvent(event: BinLogQueryEvent): Promise<void> {
+    const { query, nextPosition } = event;
+
+    // Ignore BEGIN queries
+    if (query === 'BEGIN') {
+      return;
+    }
+
+    const schemaChanges = this.toSchemaChanges(query);
+    if (schemaChanges.length > 0) {
+      this.logger.info(`Processing schema change query: ${query}`);
+      this.logger.info(`Stopping BinLog Listener to process ${schemaChanges.length} schema change events...`);
+      // Since handling the schema changes can take a long time, we need to stop the Zongji listener instead of pausing it.
+      await this.stopZongji();
+
+      for (const change of schemaChanges) {
+        await this.eventHandler.onSchemaChange(change);
+      }
+
+      // DDL queries are auto commited, and so do not come with a corresponding Xid event.
+      // This is problematic for DDL queries which result in row events, so we manually commit here.
+      this.binLogPosition.offset = nextPosition;
+      const LSN = new common.ReplicatedGTID({
+        raw_gtid: this.currentGTID!.raw,
+        position: this.binLogPosition
+      }).comparable;
+      await this.eventHandler.onCommit(LSN);
+
+      this.logger.info(`Successfully processed schema changes.`);
+      // Restart the Zongji listener
+      await this.restartZongji();
+    }
+  }
+
+  /**
+   *  Function that interprets a DDL query for any applicable schema changes.
+   *  If the query does not contain any relevant schema changes, an empty array is returned.
+   *
+   *  @param query
+   */
+  private toSchemaChanges(query: string): SchemaChange[] {
+    const ast = this.sqlParser.astify(query, { database: 'MySQL' });
```

Comment:
How closely does this parser follow the MySQL syntax? And what happens if it runs into syntax it doesn't support?

One example of a past issue:
https://github.com/taozhi8833998/node-sql-parser/issues/2232

To what extent would it be feasible to use the tablemap events instead of or in addition to this to detect changes?

Reply:
Here's an example that crashes the process with a nasty error:

```sql
CREATE TABLE sales (
    sale_date DATE,
    amount DECIMAL(10,2)
) PARTITION BY RANGE (YEAR(sale_date)) (
    PARTITION p2023 VALUES LESS THAN (2024),
    PARTITION p2024 VALUES LESS THAN (2025),
    PARTITION pmax VALUES LESS THAN MAXVALUE
);
```

The crash happens regardless of whether the table is referenced in sync rules or not, and you cannot recover by restarting - it will immediately crash at the same point again.